dbms/src/Storages/S3/FileCache.cpp

```diff
@@ -166,6 +233,72 @@ FileSegmentPtr FileCache::get(const S3::S3FilenameView & s3_fname, const std::op
     return nullptr;
 }
 
+FileSegmentPtr FileCache::getOrWait(const S3::S3FilenameView & s3_fname, const std::optional<UInt64> & filesize)
+{
+    auto s3_key = s3_fname.toFullKey();
+    auto file_type = getFileType(s3_key);
+    auto & table = tables[static_cast<UInt64>(file_type)];
+
+    std::unique_lock lock(mtx);
+
+    auto f = table.get(s3_key);
+    if (f != nullptr)
+    {
+        lock.unlock();
+        f->setLastAccessTime(std::chrono::system_clock::now());
+        auto status = f->waitForNotEmpty();
+        if (status == FileSegment::Status::Complete)
+        {
+            GET_METRIC(tiflash_storage_remote_cache, type_dtfile_hit).Increment();
+            return f;
+        }
+        else
+        {
+            // On-going download failed, let the caller retry.
+            return nullptr;
+        }
+    }
+
+    GET_METRIC(tiflash_storage_remote_cache, type_dtfile_miss).Increment();
+
+    auto estimzted_size = filesize ? *filesize : getEstimatedSizeOfFileType(file_type);
+    if (!reserveSpaceImpl(file_type, estimzted_size, /*try_evict*/ true))
+    {
+        // Space not enough.
+        GET_METRIC(tiflash_storage_remote_cache, type_dtfile_full).Increment();
+        LOG_DEBUG(
+            log,
+            "s3_key={} space not enough(capacity={} used={} estimzted_size={}), skip cache",
+            s3_key,
+            cache_capacity,
+            cache_used,
+            estimzted_size);
+
+        // Just throw, no need to let the caller retry.
+        throw Exception( //
+            ErrorCodes::S3_ERROR,
+            "Cannot reserve {} space for object {}",
+            estimzted_size,
+            s3_key);
+        return nullptr;
+    }
+
+    auto file_seg
+        = std::make_shared<FileSegment>(toLocalFilename(s3_key), FileSegment::Status::Empty, estimzted_size, file_type);
+    table.set(s3_key, file_seg);
+    lock.unlock();
+
+    PerfContext::file_cache.fg_download_from_s3++;
+    fgDownload(lock, s3_key, file_seg);
```

Comment:
The lock on `mtx` is unlocked before calling `fgDownload`. Passing it to `fgDownload` will make us feel that `fgDownload` is protected under the lock while it doesn't.
I think we should remove the param `lock` and just let `fgDownload` calls `remove(s3_key);` @Lloyd-Pottiger @breezewish

Reply:
fixed